Siegel spaces

fedelopez77 opened the pull request:
Implementation of two models of the Siegel Space.

- Upper Half space model: generalizes the upper half plane model of the hyperbolic plane.
- Bounded domain model: generalizes the Poincare ball model.

Points in these manifolds are complex symmetric matrices.

Reference: 
Federico López, Beatrice Pozzetti, Steve Trettel, Michael Strube, Anna Wienhard.
["Symmetric Spaces for Graph Embeddings: A Finsler-Riemannian Approach"](https://arxiv.org/abs/2106.04941), 2021.

ferrine commented on geoopt/manifolds/siegel/bounded_domain.py:
```diff
@@ -0,0 +1,189 @@
+
+    name = "Bounded Domain"
+
+    def __init__(self, metric: str = SiegelMetricType.RIEMANNIAN.value, rank: int = None):
```
With `Enum` you can do the following

```
from enum import Enum
class SiegelMetricType(Enum):
    """Supported metric types for Siegel Spaces"""
    RIEMANNIAN = "riem"
    FINSLER_ONE = "fone"
    FINSLER_INFINITY = "finf"
    FINSLER_MINIMUM = "fmin"
    WEIGHTED_SUM = "wsum"
```
and later
```SiegelMetricType("riem")
<SiegelMetricType.RIEMANNIAN: 'riem'>
```
or 
```
SiegelMetricType(SiegelMetricType.RIEMANNIAN)
<SiegelMetricType.RIEMANNIAN: 'riem'>
```
so no need in `from_str` method

fedelopez77 replied:
Done, thanks!

ferrine commented on geoopt/manifolds/siegel/bounded_domain.py:
```diff
@@ -0,0 +1,189 @@
+
+    name = "Bounded Domain"
+
+    def __init__(self, metric: str = SiegelMetricType.RIEMANNIAN.value, rank: int = None):
```
```suggestion
    def __init__(self, metric: str = SiegelMetricType.RIEMANNIAN, rank: int = None):
```

fedelopez77 replied:
Done

ferrine commented on geoopt/manifolds/siegel/vvd_metrics.py:
```diff
@@ -0,0 +1,181 @@
+    FINSLER_MINIMUM = "fmin"
+    WEIGHTED_SUM = "wsum"
+
+    @staticmethod
```
This method is redundant

fedelopez77 replied:
Removed, thanks!

ferrine commented on geoopt/manifolds/siegel/vvd_metrics.py:
```diff
@@ -0,0 +1,181 @@
+    @classmethod
+    def get(cls, type_str: str, rank: int):
+        metrics_map = {
+            SiegelMetricType.RIEMANNIAN.value: RiemannianMetric,
```
```suggestion
            SiegelMetricType.RIEMANNIAN: RiemannianMetric,
```
No need in `.value`

fedelopez77 replied:
Done, thanks!

ferrine commented on geoopt/manifolds/siegel/vvd_metrics.py:
```diff
@@ -0,0 +1,181 @@
+
+    @classmethod
+    def get(cls, type_str: str, rank: int):
+        metrics_map = {
```
This can be a static class variable

fedelopez77 replied:
Since I cannot have the children classes declared in a static class variable in the parent class, I extracted the creation logic in a `SiegelMetricFactory`

ferrine commented on tests/test_caley_transform.py:
```diff
@@ -0,0 +1,91 @@
+    return seed
+
+
+@pytest.fixture(scope="module", params=[torch.complex128, torch.complex64])
```
maybe add float dtype to the test suit?

fedelopez77 replied:
We use the Cayley Transform to move from one model (Upper half or Bounded domain) to the other and vice versa, thus we only define it for complex matrices.

ferrine reviewed (changes requested):
The PR looks amazing from the design to implementation. I have minor comments for code sanity and code style and no other concerns this is good. Do not forget to add an entry to CHANGELOG (and README) :)

ferrine commented:
As linters say, `make black` in the root of the repo may help)

fedelopez77 commented:
* Corrected param to init Siegel manifolds: now it is a SiegelMetricType instead of a string
 * Moved creation logic for Siegel metrics from `SiegelMetric` class to `SiegelMetricFactory`
 * Corrected tests according to these changes
 * Added entries in CHANGELOG and README
 * Run `make black` for formatting

fedelopez77 commented:
Fixed issues with pydocstyle and lint

ferrine commented:
Waiting for the tests pass now

ferrine commented:
Great! It's merged!
